Capture state changes scheduled between render and effect

dcalhoun opened the pull request:
## What?
Fixes #32154. Supersedes #32831.  Avoid stale mappings overwriting new state in specific contexts where `onStoreChange` updates the state before the effect writes the initial mapping.

## Why?
In the event that a child component dispatched an action modifying the store before the selector mapping was written, the changes in the store were lost due to `useSelect`'s `useIsomorphicLayoutEffect` referencing a stale value.

A specific example of this issue occurring is the following code where `onUnselect` is called. This call triggers an update to the store the occurs in between `useSelect`'s reading and writing the selector value, resulting in a stale `isTitleSelected` value. This resulted in the blocks inserted to the incorrect location, which is captured in https://github.com/WordPress/gutenberg/issues/32154. 

https://github.com/WordPress/gutenberg/blob/3987f612a77fd346093de1a1239aea11a6dc8e38/packages/editor/src/components/post-title/index.native.js#L36-L48

https://github.com/WordPress/gutenberg/blob/3987f612a77fd346093de1a1239aea11a6dc8e38/packages/editor/src/components/provider/use-block-editor-settings.native.js#L31

## How?
Avoid setting a reference to the selector output mapping until after the selector is run. 

## Testing Instructions
The following was manually tested: 

1. Launch the native Demo editor.
2. Focus the first block. 
3. Tap the + button to add a block.
4. **Expected:** The "ADD BLOCK HERE" insertion point is displayed beneath the selected block.
5. Focus the title for the post.
6. Tap the + button to add a block.
7. **Expected:** The "ADD BLOCK HERE" insertion point is displayed beneath the title. 
2. Focus the first block.
3. Tap the + button to add a block.
4. **Expected:** The "ADD BLOCK HERE" insertion point is displayed beneath the selected block.

## Screenshots or screencast 

<details><summary>Before</summary>

https://user-images.githubusercontent.com/438664/190246371-85c9d7bf-500a-4409-9b1b-597ac7d8eeb1.mov

</details>

<details><summary>After</summary>

https://user-images.githubusercontent.com/438664/190246407-e31b51d0-740a-4962-9d00-51882a4fb914.mov

</details>

dcalhoun committed:
Add failing useSelect test case when state updates before subscription

In the event that a child component dispatched an action modifying the
store, the changes in the store were lost due to `useSelect`'s
`useIsomorphicLayoutEffect` referencing a stale value.

dcalhoun committed:
Capture state changes before useSelect subscription

In the event that a child component dispatched an action modifying the
store, the changes in the store were lost due to `useSelect`'s
`useIsomorphicLayoutEffect` referencing a stale value.

This change ensures the `useIsomorphicLayoutEffect` references the
latest `mapOutput` when setting the related `latestMapOutput.current`
value.

dcalhoun committed:
Revert native editor workaround for stale useSelect value

Store updates triggered from within the post title component's
`componentDidUpdate` hook resulted in stale `isTitledSelected` values
returned from the memoized `useSelect`. This reverts a workaround that
avoided memoization by splitting the `useSelect` hook usage in two.

dcalhoun commented:
Hey @nerrad and @youknowriad. 👋🏻 This change touches fairly low-level logic. From what I found, you both have discussed the existing code a good bit. I would appreciate your perspective on these changes. It is not a rush, so please review at your convenience. Thank you! 🙇🏻

jsnajdr commented:
Hi @dcalhoun :wave: seems you're really up to something 🙂 The `useSelect` code does something like
```js
// read the current value
const mapOutput = latestMapOutput.current;

// indiscriminately write the read value back into the ref in an effect
useIsomorphicLayoutEffect( () => {
  latestMapOutput.current = mapOutput;
} );

// write to the ref also in subscription callback
subscribe( function onStoreChange() {
  latestMapOutput.current = newValue;
} );
```
Your test example works in such a way that the `onStoreChange` callback is called between the reading and the effect writing. We read an old value (0), then `onStoreChange` changes the value from 0 to 1 and finally the effect changes it back from 1 to 0 (!!).

> action modifying the store before the `useSelect` subscription is set,

At the moment when the action is being dispatched the `useSelect` subscription is already set: the `onStoreChange` callback gets called. But the timing of the update is such that the value is overwritten with a stale one.

jsnajdr commented:
I remembered that the `react-redux`'s `useSelector` hook uses the same approach: writing `latestSelectedState.current` in an effect on each call, no matter if the selector was run or not. Then it should have the same bug. Does it?

I rewrote the unit test that this PR is adding to Redux and observed that a stale value is really written: `useSelector` reads old value from ref, schedules an effect, executes a listener, writes a new value to the ref, and then the effect writes the stale value. But there is one crucial difference. `useSelector` stores not only the selected state in a ref, but also the "source" Redux state. And a `useSelector` calls whether that stored source state is equal to the current one, and if it isn't, it will call the selector. So, the order of events is:
1. Read old value and old state
2. Schedule a write in effect
3. Return old value from the hook
4. Listener gets called, writes new value and new state
5. Effect writes old value and old state (bad!)
6. The next call to `useSelector` reads old value and old state
7. Finds out that current state != old (stored) state
8. Calls the selector, which returns new value
9. Schedules a write in effect, which is going to write new value and new state
10. Return new value

So, the stale value is written, but it will never be used. On the other hand, `@wordpress/data` `useSelect` doesn't have a source state to compare. It will not and can not do the comparison in step 7 and call the selector. It will trust the stored value, which is wrong.

jsnajdr commented on packages/data/src/components/use-select/index.js:
```diff
@@ -191,10 +191,17 @@ export default function useSelect( mapSelect, deps ) {
+
+		latestMapOutput.current = mapOutput;
+	}, [ mapOutput ] );
+
```
A little more elegant fix would be to keep the `latestMapOutput.current` write in the original hook, but do it only if the selector was already called:
```js
let selectorRan = false;
if ( conditions ) {
  mapOutput = wrapSelect( mapSelect );
  selectorRan = true;
}

useIsomorphicLayoutEffect( () => {
  if ( selectorRan ) {
    latestMapOutput.current = mapOutput;
  }
  // ... other writes in this hook stay intact
} );
```
I tried it locally and all `useSelect` unit tests are passing. Let's see how the entire Gutenberg unit and e2e suite will do.

Also, we should the same changes in `useSuspenseSelect`. Its implementation is similar but slightly different, and we didn't unify them together yet.

jsnajdr commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -434,6 +434,67 @@ describe( 'useSelect', () => {
+		it( 'captures state changes before the subscription is set', () => {
+			registry.registerStore( 'store-1', counterStore );
+			registry.registerStore( 'store-2', counterStore );
+			registry.registerStore( 'store-3', counterStore );
```
You only need to register `store-1` here.

jsnajdr commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -434,6 +434,67 @@ describe( 'useSelect', () => {
+					( select ) =>
+						selectCount1AndDep() || {
+							count1: select( 'store-1' ).getCounter(),
+						},
```
The calls to the `mapSelect` function can be tracked more elegantly with:
```js
const selectCount1AndDep = jest.fn( ( select ) => ( {
  count1: select( 'store-1' ).getCounter(),
} ) );

/* ... */

const state = useSelect( selectCount1AndDep, [] );
```

jsnajdr commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -434,6 +434,67 @@ describe( 'useSelect', () => {
+				}
+			}
+
+			const TestComponent = jest.fn( () => {
```
We never look at how many times `TestComponent` was called or what its arguments were, so wrapping it with `jest.fn()` is not needed.

jsnajdr commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -434,6 +434,67 @@ describe( 'useSelect', () => {
+			}
+
+			const TestComponent = jest.fn( () => {
+				[ childShouldDispatch, setChildShouldDispatch ] =
```
Instead of storing a reference to `setChildShouldDispatch` and then calling it, we could achieve the same by rendering a button and triggering an event:
```js
<button onClick={ () => setChildShouldDispatch( true ) }>trigger</button>

fireEvent.click( rendered.getbyText( 'trigger' ) );
```
That's more in the spirit of React unit tests, working only with the DOM UI that any other user can also use, without accessing internals.

dcalhoun committed:
refactor: Simplify and improve useSelect test

dcalhoun committed:
refactor: Delay mapping selector output until selector is called

This avoids stale mappings overwriting new values in specific contexts
where `onStoreChange` updates the value before the effect writes the
original mapping.

dcalhoun commented:
Thank you for the thorough review and investigation, @jsnajdr. 🙇🏻 

> Your test example works in such a way that the `onStoreChange` callback is called between the reading and the effect writing. We read an old value (0), then `onStoreChange` changes the value from 0 to 1 and finally the effect changes it back from 1 to 0 (!!).
> 
> > action modifying the store before the `useSelect` subscription is set,
> 
> At the moment when the action is being dispatched the `useSelect` subscription is already set: the `onStoreChange` callback gets called. But the timing of the update is such that the value is overwritten with a stale one.

This makes sense now that you explain it so clearly. 

> So, [in `react-redux`] the stale value is written, but it will never be used. On the other hand, `@wordpress/data` `useSelect` doesn't have a source state to compare. It will not and can not do the comparison in step 7 and call the selector. It will trust the stored value, which is wrong.

This is very interesting, and very educational for me in regards to how `@wordpress/data` operates.

dcalhoun commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -434,6 +434,63 @@ describe( 'useSelect', () => {
 			);
 		} );
 
+		it( 'captures state changes before the subscription is set', () => {
```
@jsnajdr I am struggling to identify a concise way to describe this test based upon your findings. What do you think is an appropriate description for this test?

jsnajdr replied:
I would say something like "captures state changes scheduled between render and effect." The key part of `useSelect` here is that it reads a ref in render and writes it in effect:
```js
function useSelect() {
  const value = ref.current;
  useEffect( () => {
    ref.current = value;
  } );
}
```
And then there is some `Inner` component that writes the ref in effect, too:
```js
function Inner() {
  useEffect( () => {
    ref.current = newValue;
  } );
  return null;
}
```
Now when you combine them together:
```js
function Outer() {
  useSelect();
  return <Inner />;
}
```
the effects are executed in children first and in parents second, so the order of operations is:
1. outer render reads
2. inner effect writes new value
3. outer effect writes stale value

In reality the `useSelect` effects are layout effects and the `Inner` component uses `componentDidUpdate`, but it leads to the same scheduling.

jsnajdr commented on packages/data/src/components/use-select/index.js:
```diff
@@ -188,20 +190,16 @@ export default function useSelect( mapSelect, deps ) {
 
+		if ( selectorRan ) {
+			latestMapOutput.current = mapOutput;
+		}
```
It's probably just for aesthetics sake, but I'd rather keep the assignments to `latestMapOutput` and `latestMapOutputError` close to each other.

jsnajdr reviewed (approved):
Looks good :+1:

dcalhoun committed:
fix: Rename test description

This test focuses upon the order in which state is read and written. It
is not related to when the subscription is set.

dcalhoun committed:
refactor: Group related selector output mapping logic

Reduce cognitive load of code by grouping related logic.
